x/merkledb/benchmarks/benchmark.go

```diff
@@ -0,0 +1,372 @@
+		return err
+	}
+	return nil
+}
+
+func runBenchmark() error {
+	promRegistry := prometheus.NewRegistry()
+	if err := prometheus.Register(promRegistry); err != nil {
+		return err
+	}
+
+	writesCounter := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "merkledb_bench",
+		Name:      "writes",
+		Help:      "Total number of keys written",
+	})
+	batchCounter := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "merkledb_bench",
+		Name:      "batch",
+		Help:      "Total number of batches written",
+	})
+	deleteRate := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "entry_delete_rate",
+		Help:      "The rate at which elements are deleted",
+	})
+	updateRate := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "entry_update_rate",
+		Help:      "The rate at which elements are updated",
+	})
+	insertRate := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "entry_insert_rate",
+		Help:      "The rate at which elements are inserted",
+	})
+	batchWriteRate := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "batch_write_rate",
+		Help:      "The rate at which the batch was written",
+	})
+	promRegistry.MustRegister(writesCounter)
+	promRegistry.MustRegister(batchCounter)
+	promRegistry.MustRegister(deleteRate)
+	promRegistry.MustRegister(updateRate)
+	promRegistry.MustRegister(insertRate)
+	promRegistry.MustRegister(batchWriteRate)
+
+	http.Handle("/metrics", promhttp.Handler())
+
+	server := &http.Server{
+		Addr:              fmt.Sprintf(":%d", *httpMetricPort),
+		ReadHeaderTimeout: 3 * time.Second,
+	}
+	go func() {
+		err := server.ListenAndServe()
+		if err != nil && err != http.ErrServerClosed {
+			fmt.Fprintf(os.Stderr, "unable to listen and serve : %v\n", err)
+		}
+	}()
+
+	levelDB, err := leveldb.New(
+		getRunningDatabaseDirectory(),
+		nil,
+		logging.NoLog{},
+		promRegistry,
+	)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create levelDB database : %v\n", err)
+		return err
+	}
+
+	mdb, err := merkledb.New(context.Background(), levelDB, getMerkleDBConfig(promRegistry))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create merkleDB database : %v\n", err)
+		return err
+	}
+	defer func() {
+		mdb.Close()
+		levelDB.Close()
+	}()
+
+	low := uint64(0)
+	var deleteDuration, addDuration, updateDuration, batchDuration time.Duration
+	batch := mdb.NewBatch()
+	for {
+		startBatchTime := time.Now()
+
+		// delete first 2.5k keys from the beginning
+		startDeleteTime := time.Now()
+		for keyToDeleteIdx := low; keyToDeleteIdx < low+databaseRunningBatchSize; keyToDeleteIdx++ {
+			entryHash := calculateIndexEncoding(keyToDeleteIdx)
+			err = batch.Delete(entryHash)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "unable to delete merkleDB entry : %v\n", err)
+				return err
+			}
+		}
+		deleteDuration = time.Since(startDeleteTime)
```

Comment:
You can avoid all this math and just report the raw number of deletes. Grafana can convert this to a rate for you.

Reply:
I've added both. I believe that my calculation would be more accurate, but let's have both for the time being.